### ISSUE TITLE
std-instead-of-alloc: there is no `alloc::collections::VecDeque`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16695]
Using the following flags

`--force-warn clippy::std-instead-of-alloc`

this code:

```rust
#![warn(clippy::all, clippy::pedantic, clippy::nursery, clippy::restriction)]
#![allow(clippy::missing_docs_in_private_items, clippy::blanket_clippy_restriction_lints)]
use std::collections::VecDeque;

fn main() {
    let mut c: VecDeque<u8> = VecDeque::new();
    c.swap(0, 1);
    c.push_back(1);

    let _ = c[0]; // clippy::indexing_slicing correctly detects possible panic

    c.swap(0, 1); // Could cause panic; clippy ignores this line
    // c.swap(1, 2); // Example swap panic
}
```

caused the following diagnostics:

```
    Checking _b4e4b49c96533ac40e3e725c4362a6857c3af18f v0.1.0 (/tmp/icemaker_global_tempdir.gPgWiqot4T9Z/icemaker_clippyfix_tempdir.yqcDcwegV0PK/_b4e4b49c96533ac40e3e725c4362a6857c3af18f)
warning: used import from `std` instead of `alloc`
 --> src/main.rs:3:5
  |
3 | use std::collections::VecDeque;
  |     ^^^ help: consider importing the item from `alloc`: `alloc`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#std_instead_of_alloc
  = note: requested on the command line with `--force-warn clippy::std-instead-of-alloc`

warning: `_b4e4b49c96533ac40e3e725c4362a6857c3af18f` (bin "_b4e4b49c96533ac40e3e725c4362a6857c3af18f") generated 1 warning (run `cargo clippy --fix --bin "_b4e4b49c96533ac40e3e725c4362a6857c3af18f" -p _b4e4b49c96533ac40e3e725c4362a6857c3af18f -- -Aclippy::all -Awarnings --force-warn clippy::std-instead-of-alloc --cap-lints warn` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s

```

However after applying these diagnostics, the resulting code:

```rust
#![warn(clippy::all, clippy::pedantic, clippy::nursery, clippy::restriction)]
#![allow(clippy::missing_docs_in_private_items, clippy::blanket_clippy_restriction_lints)]
use alloc::collections::VecDeque;

fn main() {
    let mut c: VecDeque<u8> = VecDeque::new();
    c.swap(0, 1);
    c.push_back(1);

    let _ = c[0]; // clippy::indexing_slicing correctly detects possible panic

    c.swap(0, 1); // Could cause panic; clippy ignores this line
    // c.swap(1, 2); // Example swap panic
}
```
no longer compiled:

```
    Checking _b4e4b49c96533ac40e3e725c4362a6857c3af18f v0.1.0 (/tmp/icemaker_global_tempdir.gPgWiqot4T9Z/icemaker_clippyfix_tempdir.yqcDcwegV0PK/_b4e4b49c96533ac40e3e725c4362a6857c3af18f)
error[E0433]: cannot find module or crate `alloc` in this scope
 --> src/main.rs:3:5
  |
3 | use alloc::collections::VecDeque;
  |     ^^^^^ use of unresolved module or unlinked crate `alloc`
  |
  = help: add `extern crate alloc` to use the `alloc` crate

For more information about this error, try `rustc --explain E0433`.
error: could not compile `_b4e4b49c96533ac40e3e725c4362a6857c3af18f` (bin "_b4e4b49c96533ac40e3e725c4362a6857c3af18f") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_b4e4b49c96533ac40e3e725c4362a6857c3af18f` (bin "_b4e4b49c96533ac40e3e725c4362a6857c3af18f" test) due to 1 previous error

```

Version:
```
rustc 1.96.0-nightly (e3d66fe39 2026-03-07)
binary: rustc
commit-hash: e3d66fe39ae70380fa2365c008e2927479114844
commit-date: 2026-03-07
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.0

```

**Comment by PatchMixolydic** [comment]
[`alloc::collections::VecDeque` exists](https://doc.rust-lang.org/stable/alloc/collections/vec_deque/struct.VecDeque.html) ([reexport here](https://doc.rust-lang.org/stable/alloc/collections/index.html#:~:text=pub%20use%20vec%5Fdeque%3A%3AVecDeque%3B)). The issue here is that unlike `core` and `std`, [`alloc` is never automatically added to the extern prelude](https://doc.rust-lang.org/reference/names/preludes.html#r-names.preludes.extern.edition2018:~:text=Additional%20crates%20that%20ship%20with%20rustc), rendering it inaccessible by default. The fix is already suggested by E0433:

```
  = help: add `extern crate alloc` to use the `alloc` crate
```

Note that this applies on all editions, even on Rust 2018 and newer.

It would be nice if this issue was detected by `clippy::std_instead_of_alloc`, but I imagine most users of this lint would already be aware of the need for `extern crate alloc;`.

**Comment by bushrat011899** [comment]
I think this can be considered a duplicate of #15836, since they're caused by the same fundamental issue of not checking what items are in scope before making a suggestion.
